ground/src/main/java/com/google/android/ground/persistence/sync/MediaUploadWorker.kt

```diff
@@ -0,0 +1,184 @@
+          it.copy(
+            syncStatus = Mutation.SyncStatus.MEDIA_UPLOAD_IN_PROGRESS,
+            retryCount = it.retryCount + 1
+          )
+        }
+    mutationRepository.updateMutations(mutations) // Mark MEDIA_UPLOAD_IN_PROGRESS
+
+    val mutationStatusesAfterUpload = uploadMediaAndGetStatus(mutations)
+
+    // The worker proceeds to a fixed point as follows:
+    //   If there are any FAILED or PENDING mutations after the worker operation, we retry.
+    //   During the retry, FAILED mutations will not be processed, while PENDING ones will only be
+    //   reattempted if they satisfy the retry count bound. Eventually, either all of the
+    //   mutations for this LOI will have succeeded, failed, or exceeded the retry count.
+    if (
+      mutationStatusesAfterUpload.any {
+        it == Mutation.SyncStatus.FAILED || it == Mutation.SyncStatus.MEDIA_UPLOAD_PENDING
+      }
+    ) {
+      return Result.retry()
+    }
+    return Result.success()
+  }
+
+  /**
+   * Attempts to upload associated media for a given list of [SubmissionMutation] and updates the
+   * status of each mutation based on whether uploads succeeded or failed.
+   */
+  private suspend fun uploadMediaAndGetStatus(
+    mutations: List<SubmissionMutation>
+  ): List<Mutation.SyncStatus> =
+    mutations
+      .map { uploadSubmissionMediaAndUpdateMutation(it) }
+      .also { resultingMutations -> mutationRepository.updateMutations(resultingMutations) }
+      .map { it.syncStatus }
+
+  /**
+   * Attempts to upload all media associated with a given submission. Updates the submission's sync
+   * status depending on whether or the uploads failed or succeeded.
+   */
+  private suspend fun uploadSubmissionMediaAndUpdateMutation(
+    mutation: SubmissionMutation
+  ): SubmissionMutation =
+    uploadSubmissionPhotos(mutation)
+      .fold(
+        onSuccess = { mutation.copy(syncStatus = Mutation.SyncStatus.COMPLETED) },
+        onFailure = {
+          val cause = it.message ?: "unknown upload error"
+          if (it is FileNotFoundException) {
+            return mutation.copy(syncStatus = Mutation.SyncStatus.FAILED, lastError = cause)
+          }
+
+          return mutation.copy(
+            syncStatus = Mutation.SyncStatus.MEDIA_UPLOAD_PENDING,
+            lastError = cause
+          )
+        }
+      )
+
+  private suspend fun uploadSubmissionPhotos(mutation: SubmissionMutation): kotlin.Result<Unit> =
+    // TODO: Use media response types instead of discriminating on Task.Type.
+    // For example, we should pass a List<PhotoResponse> to uploadPhotoMedia(), which can take care
+    // of the bulk of the response-specific work.
+    mutation.deltas
+      .filter { (_, taskType, newValue): ValueDelta ->
+        taskType === Task.Type.PHOTO && newValue.isNotNullOrEmpty()
+      }
+      .map { (_, _, newValue): ValueDelta -> newValue.toString() }
+      .map { uploadPhotoMedia(it) }
+      .reduce { a, b -> combineResults(a, b) }
```

Comment:
If one photo in a submission fails, will it try to upload all photos again, or does each photo have a separate mutation with its own state?

Also, I needed to understand `reduce` and to look at the definition of fun `combineResults` to know what this does. Is it just returning the first exception? If we don't do anything with it, consider returning a `boolean` instead.

Reply:
At the moment, it will attempt to re-upload all the photos in the response. This is something we'll have to address. This is part of what drove me to use an initial approach of having a separate mutation type for the media itself (we could have one for each photo). We can probably still avoid having to do that, but we will need to write the extra bit somewhere to track individual statuses on each photo. Once we use `PhotoResponse` here, we'll have to likely set a flag on that value and update the submission mutation to indicate success/failure of each photo.

re: reduce/combine. Yes, this walks through a list of Results, which each may either be an exception or success. If the left argument is an exception, that just gets returned and propagated all the way through so `[FileNotFound Success OtherError Success]` would return `FileNotFound`. Only all successes will return success. 

We do use the exception to determine whether or not the failure might be recoverable and thus the upload worthy of retry. Right now, we only check for `FileNotFound` in which case we explicitly mark it as failed and don't bother retrying the upload. We currently treat all other failures as transient and potential candidates for retry.

Reply:
Could you please add a TODO and bug to track retry of individual photos?